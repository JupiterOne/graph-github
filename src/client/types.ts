export type ResourceIteratee<T> = (each: T) => Promise<void> | void;
