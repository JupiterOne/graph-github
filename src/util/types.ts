export interface JsonWebToken {
  exp: number;
  iat: number;
  iss: any;
}

export interface SignedJsonWebToken extends JsonWebToken {
  token: string;
}
