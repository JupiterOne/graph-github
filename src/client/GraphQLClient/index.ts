export * from './types/client';
export * from './client';
