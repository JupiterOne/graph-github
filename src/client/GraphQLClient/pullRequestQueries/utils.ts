import { URL } from 'url';

const innerResourcePaginationRequired = (pullRequest): boolean => {
  if (!pullRequest) {
    return false;
  }
  return (
    pullRequest.commits?.pageInfo.hasNextPage ||
    pullRequest.reviews?.pageInfo.hasNextPage ||
    pullRequest.labels?.pageInfo.hasNextPage
  );
};

const responseToResource = (node) => {
  return {
    ...node,
    commits: node.commits?.nodes?.map((node) => node.commit) ?? [],
    reviews: node.reviews?.nodes ?? [],
    labels: node.labels?.nodes ?? [],
  };
};

const hasRepoOwnerAndName = (node) => {
  const repoOwnerAndName = determineRepoOwnerAndName(node);
  return repoOwnerAndName.repoOwner && repoOwnerAndName.repoName;
};

const determineRepoOwnerAndName = (
  node,
): { repoOwner?: string; repoName?: string } => {
  const urlPath = node.url // ex: https://github.com/JupiterOne/graph-github/pull/1
    ? new URL(node.url)?.pathname // ex: /JupiterOne/graph-github/pull/4"
    : '';

  // Attempt to pull repo name and owner from graphQL response. If not there, parse the pull request url.
  const repoOwner = node.headRepository?.owner?.login ?? urlPath.split('/')[1]; // ex: JupiterOne
  const repoName = node.headRepository?.name ?? urlPath.split('/')[2]; // ex: graph-github

  return {
    repoOwner,
    repoName,
  };
};

export default {
  innerResourcePaginationRequired,
  responseToResource,
  findRepoOwnerAndName: determineRepoOwnerAndName,
  hasRepoOwnerAndName,
};
