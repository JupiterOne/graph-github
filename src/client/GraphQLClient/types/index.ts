export * from './client';
export * from './state';
