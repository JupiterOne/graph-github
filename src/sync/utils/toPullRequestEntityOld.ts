import {
  parseTimePropertyValue,
  setRawData,
} from '@jupiterone/integration-sdk-core';
import { omit, uniq } from 'lodash';
import { GITHUB_PR_ENTITY_TYPE, GITHUB_PR_ENTITY_CLASS } from '../../constants';
import {
  PullsListResponseItem,
  PullsListCommitsResponseItem,
  IdEntityMap,
  UserEntity,
  PullRequestEntity,
} from '../../types';
import {
  flattenMatrix,
  aggregateProperties,
  displayNamesFromLogins,
} from '../../util/propertyHelpers';
import { Approval } from '../converters';

export function toPullRequestEntityOld(
  data: PullsListResponseItem,
  commits?: PullsListCommitsResponseItem[],
  commitsApproved?: PullsListCommitsResponseItem[],
  commitsByUnknownAuthor?: PullsListCommitsResponseItem[],
  approvals?: Approval[],
  usersByLogin?: IdEntityMap<UserEntity>,
): PullRequestEntity {
  const commitHashes = commits ? commits.map((c) => c.sha) : undefined;
  const commitMessages = commits
    ? commits.map((c) => c.commit.message)
    : undefined;
  const commitsApprovedHashes = commitsApproved
    ? commitsApproved.map((c) => c.sha)
    : undefined;
  const commitsByUnknownAuthorHashes = commitsByUnknownAuthor
    ? commitsByUnknownAuthor.map((c) => c.sha)
    : undefined;
  const commitsNotApproved = commitHashes
    ? commitHashes.filter((c) => !commitsApprovedHashes!.includes(c))
    : undefined;

  const approved = commitsNotApproved
    ? commitsNotApproved.length === 0
    : undefined;
  const validated = commitsByUnknownAuthor
    ? commitsByUnknownAuthor.length === 0
    : undefined;

  const approverLogins = approvals
    ? uniq(
        flattenMatrix<string>(
          aggregateProperties<string[]>('approverUsernames', approvals),
        ),
      )
    : undefined;
  const userLogin: string = data.user ? data.user.login : '';
  const authorUser = (usersByLogin || {})[userLogin];
  let reviewerLogins: string[] = [];
  let reviewers: string[] = [];
  if (data.requested_reviewers) {
    reviewerLogins = aggregateProperties<string>(
      'login',
      data.requested_reviewers,
    );
    reviewers = data.requested_reviewers.reduce(
      (reviewers: string[], reviewerData) => {
        if (reviewerData) {
          if (usersByLogin && usersByLogin[reviewerData.login]) {
            reviewers.push(usersByLogin[reviewerData.login].displayName!);
          } else {
            reviewers.push(reviewerData.login);
          }
        }
        return reviewers;
      },
      [],
    );
  }
  const entity: PullRequestEntity = {
    _type: GITHUB_PR_ENTITY_TYPE,
    _class: [GITHUB_PR_ENTITY_CLASS],
    _key: `${data.base.repo.full_name}/pull-requests/${data.number}`,
    displayName: `${data.base.repo.name}/${data.number}`,
    accountLogin: data.base.repo.owner ? data.base.repo.owner.login : '',
    repository: data.base.repo.name,
    // the type is hacked here because typing of data properties is controlled by a library call
    // so I can't just say that data.number is a string
    // here would be another way to solve it:
    // id: JSON.stringify(data.number).replace(/\"/g, ''),
    id: <string>(<unknown>data.number),

    name: data.title,
    title: data.title,
    description:
      data.body && data.body.length > 0
        ? `${data.body.substring(0, 80)}...`
        : undefined,
    webLink: data.html_url,

    state: data.state,
    open: data.state === 'open',
    mergeCommitHash: data.merge_commit_sha,
    merged: (data.merged_at as any) !== null,
    declined: data.state === 'closed' && (data.merged_at as any) === null,
    approved,
    validated,

    commits: commitHashes,
    commitMessages,
    commitsApproved: commitsApprovedHashes,
    commitsNotApproved,
    commitsByUnknownAuthor: commitsByUnknownAuthorHashes,

    source: data.head.ref,
    target: data.base.ref,

    createdOn: parseTimePropertyValue(data.created_at),
    updatedOn: parseTimePropertyValue(data.updated_at),
    mergedOn: parseTimePropertyValue(data.merged_at),

    authorLogin: userLogin,
    author: authorUser
      ? // We know displayName is set; see toOrganizationMemberEntity
        authorUser.displayName!
      : // Fallback to username. This will always be used when ingesting from a
        // user account, since we don't ingest team members in that case (there
        // is no team)
        userLogin,
    reviewerLogins: reviewerLogins,
    reviewers: reviewers,
    approverLogins,
    approvers:
      approverLogins && usersByLogin
        ? displayNamesFromLogins(approverLogins, usersByLogin)
        : approverLogins,
  };
  const rawDataPropertiesToRemove = ['head', 'base']; // a few particularly large pieces of data that are repeated on every PR
  setRawData(entity, {
    name: 'default',
    rawData: omit(data, rawDataPropertiesToRemove),
  });
  return entity;
}
