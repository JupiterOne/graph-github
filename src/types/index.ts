export * from './integration';
export * from './client';
export * from './entities';
