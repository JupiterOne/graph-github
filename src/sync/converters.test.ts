import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import {
  toOrganizationMemberEntity,
  toOrganizationCollaboratorEntity,
  toRepositoryEntity,
  toAccountEntity,
  toPullRequestEntity,
} from './converters';
import { UserEntity, PullsListResponseItem } from '../types';
import { omit } from 'lodash';
import { PullRequest } from '../client/GraphQLClient/types';
import {
  fixturePullRequest,
  fixtureUser,
  fixtureReviewerUser,
  pullRequestReviewer,
  pullRequestUser,
} from './fixtures/pullRequest';
import { toPullRequestEntityOld } from './utils/toPullRequestEntityOld';

describe('toAccountEntity', () => {
  const apiResponse = {
    id: 'account-node-id',
    login: 'account-login',
    name: 'account-name',
    createdAt: '2021-05-27T15:21:12Z',
    updatedAt: '2021-09-16T15:04:20Z',
    description: "Here's my test description",
    email: 'email@email.com',
    databaseId: 12345678,
    isVerified: false,
    location: 'Transylvania',
    websiteUrl: 'www.goclickatesturl.com',
    url: 'https://github.com/SomeOrg',
  };

  test('properties transferred', () => {
    const entity = toAccountEntity(apiResponse as any);
    expect(entity).toEqual({
      _class: ['Account'],
      _type: 'github_account',
      _key: 'account-node-id',
      _rawData: [
        {
          name: 'default',
          rawData: apiResponse,
        },
      ],
      accountType: 'Organization',
      accountId: 'account-login',
      login: 'account-login',
      name: 'account-name',
      displayName: 'account-name',
      createdOn: 1622128872000,
      updatedOn: 1631804660000,
      description: "Here's my test description",
      email: 'email@email.com',
      node: 'account-node-id',
      databaseId: 12345678,
      isVerified: false,
      location: 'Transylvania',
      websiteUrl: 'www.goclickatesturl.com',
      webLink: 'https://github.com/SomeOrg',
    });
  });

  test('displayName falls back to login', () => {
    expect(
      toAccountEntity({ ...apiResponse, name: undefined } as any),
    ).toMatchObject({
      displayName: 'account-login',
    });
  });
});

test('toRepositoryEntities', () => {
  const apiResponse = {
    id: 'repo-node-id',
    name: 'repo-name',
    nameWithOwner: 'owner/repo-name',
    isPrivate: true,
    isArchived: false,
    url: 'web-link',
  };
  const entity = toRepositoryEntity(apiResponse as any);
  expect(entity).toEqual({
    _key: 'repo-node-id',
    _type: 'github_repo',
    _class: ['CodeRepo'],
    _rawData: [
      {
        name: 'default',
        rawData: apiResponse,
      },
    ],
    webLink: 'web-link',
    public: false,
    name: 'repo-name',
    displayName: 'repo-name',
    fullName: 'owner/repo-name',
    owner: 'owner',
    archived: false,
  });
});

describe('toOrganizationMemberEntity', () => {
  const apiResponse = {
    id: 'member-node-id',
    login: 'user-login',
    name: 'User Flynn',
    hasTwoFactorEnabled: true,
    role: 'Maintainer',
    isSiteAdmin: false,
  };

  test('properties transferred', () => {
    const entity = toOrganizationMemberEntity(apiResponse as any);
    expect(entity).toEqual({
      _key: 'member-node-id',
      _type: 'github_user',
      _class: ['User'],
      _rawData: [
        {
          name: 'default',
          rawData: apiResponse,
        },
      ],
      login: 'user-login',
      username: 'user-login',
      displayName: 'User Flynn',
      name: 'User Flynn',
      mfaEnabled: true,
      role: 'Maintainer',
      siteAdmin: false,
      webLink: 'https://github.com/user-login',
    });
  });

  test('displayName falls back to login', () => {
    expect(
      toOrganizationMemberEntity({ ...apiResponse, name: undefined } as any),
    ).toMatchObject({
      displayName: 'user-login',
    });
  });
});

describe('toOrganizationCollaboratorEntity', () => {
  const apiResponse = {
    node_id: 'member-node-id',
    login: 'user-login',
    name: 'User Flynn',
  };

  test('properties transferred', () => {
    const entity = toOrganizationCollaboratorEntity(apiResponse as any);
    expect(entity).toEqual({
      _key: 'member-node-id',
      _type: 'github_user',
      _class: ['User'],
      _rawData: [
        {
          name: 'default',
          rawData: apiResponse,
        },
      ],
      login: 'user-login',
      username: 'user-login',
      displayName: 'User Flynn',
      name: 'User Flynn',
      mfaEnabled: undefined,
      role: 'OUTSIDE',
      siteAdmin: false,
      webLink: 'https://github.com/user-login',
    });
  });

  test('displayName falls back to login', () => {
    expect(
      toOrganizationMemberEntity({ ...apiResponse, name: undefined } as any),
    ).toMatchObject({
      displayName: 'user-login',
    });
  });
});

describe('toPullRequestEntity compared against toPullRequestEntityOld', () => {
  const user = ({
    name: 'Some Body',
    displayName: 'Some Body',
    login: 'somebody',
    username: 'somebody',
  } as unknown) as UserEntity;
  const reviewerUser = ({
    name: 'Reviewer User',
    displayName: 'Reviewer User',
    login: 'reviewer-user',
    username: 'reviewer-user',
  } as unknown) as UserEntity;
  const oldApiResponse: PullsListResponseItem = ({
    title: 'The Best PR Ever',
    number: '420',
    user: {
      login: 'somebody',
    },
    requested_reviewers: [
      {
        login: 'reviewer-user',
      },
    ],
    base: {
      ref: 'qwerty098765',
      repo: {
        full_name: 'my-team/my-repo',
        name: 'my-repo',
        owner: {
          login: 'my-team',
        },
      },
    },
    body: 'This is the description',
    state: 'MERGED', // Merged because it hase a merged_at property
    head: {
      ref: 'abcdef123456',
    },
    created_at: '2001-09-11T08:46:00Z',
    updated_at: '2001-09-11T20:30:00Z',
    merged_at: '2021-07-21T14:06:13Z',
    merge_commit_sha: 'f8d8a228a6046ead812d9ea0c457e429342a89f7',
    url: 'https://api.github.com/repos/JupiterOne/jupiter-project-repo/pulls/1',
    html_url: 'https://github.com/JupiterOne/jupiter-project-repo/pulls/1',
  } as unknown) as PullsListResponseItem;

  const newApiResponse: PullRequest = {
    id: 'id',
    additions: 1,
    author: pullRequestUser,
    authorAssociation: 'authorAssociation',
    baseRefName: 'qwerty098765',
    baseRefOid: 'baseRefOid',
    baseRepository: {
      name: 'my-repo',
      owner: {
        login: 'my-team',
        id: '12345',
        url: 'https://github.com/my-team',
      },
    },
    body: 'This is the description',
    changedFiles: 1,
    checksUrl: 'checksUrl',
    closed: false,
    closedAt: undefined,
    createdAt: '2001-09-11T08:46:00Z',
    deletions: 1,
    editor: pullRequestUser,
    headRefName: 'abcdef123456',
    headRefOid: 'headRefOid',
    headRepository: {
      name: 'their-repo',
      owner: {
        login: 'their-team',
        id: '67890',
        url: 'https://github.com/their-team',
      },
    },
    isDraft: false,
    lastEditedAt: undefined,
    locked: false,
    mergeCommit: {
      oid: 'f8d8a228a6046ead812d9ea0c457e429342a89f7',
      id: 'id',
      message: 'message',
      authoredDate: '2001-09-11T20:30:00Z',
      changedFiles: 1,
      commitUrl: 'commitUrl',
      author: { user: pullRequestUser },
    },
    mergeable: 'MERGEABLE',
    merged: true,
    mergedAt: '2021-07-21T14:06:13Z',
    mergedBy: pullRequestUser,
    number: 420,
    permalink: 'permalink',
    publishedAt: undefined,
    reviewDecision: 'REVIEW_REQUIRED',
    state: 'MERGED',
    title: 'The Best PR Ever',
    updatedAt: '2001-09-11T20:30:00Z',
    url: 'https://github.com/JupiterOne/jupiter-project-repo/pulls/1',
    // Optional extra traversals
    commits: undefined,
    labels: undefined,
    reviews: undefined,
  };

  const expectedEntity = {
    _type: 'github_pullrequest',
    _class: ['PR'],
    _key: `my-team/my-repo/pull-requests/420`,
    name: 'The Best PR Ever',
    displayName: `my-repo/420`,
    id: '420',
    number: 420,
    pullRequestId: 'id',
    accountLogin: 'my-team',
    authorLogin: 'somebody',
    author: 'Some Body',
    reviewerLogins: ['reviewer-user'],
    reviewers: ['Reviewer User'],
    approverLogins: ['reviewer-user'],
    approvers: ['Reviewer User'],
    title: 'The Best PR Ever',
    description: `${newApiResponse.body?.substring(0, 80)}...`,
    state: 'MERGED',
    source: 'abcdef123456',
    target: 'qwerty098765',
    repository: 'my-repo',
    createdOn: parseTimePropertyValue(newApiResponse.createdAt),
    updatedOn: parseTimePropertyValue(newApiResponse.updatedAt),
    webLink: 'https://github.com/JupiterOne/jupiter-project-repo/pulls/1',
    commits: ['commit_a', 'commit_b', 'commit_c'],
    commitMessages: ['Commit A', 'Commit B', 'Commit C'],
    commitsApproved: ['commit_a', 'commit_b'],
    commitsNotApproved: ['commit_c'],
    commitsByUnknownAuthor: [],
    approved: false,
    allCommitsApproved: false,
    validated: true,
    open: false,
    merged: true,
    declined: false,
    mergedOn: 1626876373000,
    mergeCommitHash: 'f8d8a228a6046ead812d9ea0c457e429342a89f7',
  };

  test('direct comparison', () => {
    const oldEntity = toPullRequestEntityOld(
      oldApiResponse,
      [
        { sha: 'commit_a', commit: { message: 'Commit A' } } as any,
        { sha: 'commit_b', commit: { message: 'Commit B' } } as any,
        { sha: 'commit_c', commit: { message: 'Commit C' } } as any,
      ],
      [{ sha: 'commit_a' } as any, { sha: 'commit_b' } as any],
      [],
      [{ approverUsernames: ['reviewer-user'] } as any],
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
    );
    const newEntity = toPullRequestEntity(
      {
        ...newApiResponse,
        commits: [
          { oid: 'commit_a', message: 'Commit A', author: { user } } as any,
          { oid: 'commit_b', message: 'Commit B', author: { user } } as any,
          { oid: 'commit_c', message: 'Commit C', author: { user } } as any,
        ],
        reviews: [
          {
            id: 'id',
            state: 'APPROVED',
            updatedAt: 'updatedAt',
            url: 'url',
            author: pullRequestReviewer,
            commit: {
              oid: 'commit_b',
            },
          },
        ],
      },
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
    );

    // Compare against the previous converter
    const newPullRequestProperties = [
      'allCommitsApproved',
      'pullRequestId',
      'number',
    ];
    expect(omit(newEntity, [...newPullRequestProperties, '_rawData'])).toEqual(
      omit(oldEntity, ['_rawData']),
    );

    // Compare with expected
    expect(omit(newEntity, ['_rawData'])).toEqual(expectedEntity);
  });

  test('declined comparison', () => {
    const apiResponseDeclined: PullRequest = {
      ...newApiResponse,
      state: 'CLOSED',
      merged: false,
      mergedAt: undefined,
      mergeCommit: undefined,
    };
    const newEntity = toPullRequestEntity(
      {
        ...apiResponseDeclined,
        commits: [
          { oid: 'commit_a', message: 'Commit A', author: { user } } as any,
          { oid: 'commit_b', message: 'Commit B', author: { user } } as any,
          { oid: 'commit_c', message: 'Commit C', author: { user } } as any,
        ],
        reviews: [
          {
            id: 'id',
            state: 'APPROVED',
            updatedAt: 'updatedAt',
            url: 'url',
            author: pullRequestReviewer,
            commit: {
              oid: 'commit_b',
            },
          },
        ],
      },
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
    );

    expect(omit(newEntity, ['_rawData'])).toEqual({
      ...expectedEntity,
      state: 'CLOSED',
      open: false,
      declined: true,
      merged: false,
      mergedOn: undefined,
      mergeCommitHash: undefined,
    });
  });

  test('without commits or reviews', () => {
    //TODO: fix this
    const entity = toPullRequestEntity(
      newApiResponse,
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
      {
        [user.login]: user,
        [reviewerUser.login]: reviewerUser,
      },
    );
    expect(omit(entity, ['_rawData'])).toEqual({
      ...expectedEntity,
      allCommitsApproved: undefined,
      validated: undefined,
      commits: undefined,
      commitMessages: undefined,
      commitsApproved: undefined,
      commitsNotApproved: undefined,
      commitsByUnknownAuthor: undefined,
      approvers: undefined,
      approverLogins: undefined,
      reviewers: undefined,
      reviewerLogins: undefined,
    });
  });
});

describe('toPullRequestEntity', () => {
  test('with all commits approved', () => {
    const entity = toPullRequestEntity(
      fixturePullRequest,
      {
        [fixtureUser.login!]: fixtureUser,
        [fixtureReviewerUser.login!]: fixtureReviewerUser,
      },
      {
        [fixtureUser.login!]: fixtureUser,
        [fixtureReviewerUser.login!]: fixtureReviewerUser,
      },
    );
    expect(entity).toMatchSnapshot();
  });

  test('declined', () => {
    const declinedPullRequest: PullRequest = {
      ...fixturePullRequest,
      state: 'CLOSED',
      merged: false,
      mergedAt: undefined,
      mergeCommit: undefined,
    };
    const entity = toPullRequestEntity(
      declinedPullRequest,
      {
        [fixtureUser.login]: fixtureUser as UserEntity,
        [fixtureReviewerUser.login]: fixtureReviewerUser as UserEntity,
      },
      {
        [fixtureUser.login!]: fixtureUser,
        [fixtureReviewerUser.login!]: fixtureReviewerUser,
      },
    );
    expect(entity).toMatchObject({
      state: 'CLOSED',
      open: false,
      declined: true,
      merged: false,
      mergedOn: undefined,
      mergeCommitHash: undefined,
    });
  });

  test('with no known user approvals', () => {
    const entity = toPullRequestEntity(fixturePullRequest, {}, {});
    expect(entity).toMatchObject({
      allCommitsApproved: false,
      validated: false,
      commits: fixturePullRequest.commits?.map((c) => c.oid),
      commitMessages: fixturePullRequest.commits?.map((c) => c.message),
      commitsApproved: [],
      commitsNotApproved: fixturePullRequest.commits?.map((c) => c.oid),
      commitsByUnknownAuthor: fixturePullRequest.commits?.map((c) => c.oid),
      approvers: [fixtureReviewerUser.displayName],
      approverLogins: [fixtureReviewerUser.login],
      author: fixtureUser.displayName,
      reviewers: [fixtureReviewerUser.displayName],
      _rawData: [
        {
          name: 'default',
          rawData: fixturePullRequest,
        },
      ],
    });
  });

  test('with commits and reviews from users that are contributors but not team members', () => {
    const entity = toPullRequestEntity(
      fixturePullRequest,
      {},
      {
        [fixtureUser.login!]: fixtureUser,
        [fixtureReviewerUser.login!]: fixtureReviewerUser,
      },
    );
    expect(entity).toMatchObject({
      allCommitsApproved: false,
      validated: true,
      commits: fixturePullRequest.commits?.map((c) => c.oid),
      commitMessages: fixturePullRequest.commits?.map((c) => c.message),
      commitsApproved: [],
      commitsNotApproved: fixturePullRequest.commits?.map((c) => c.oid),
      commitsByUnknownAuthor: [],
      approvers: [fixtureReviewerUser.displayName],
      approverLogins: [fixtureReviewerUser.login],
      author: fixtureUser.displayName,
      reviewers: [fixtureReviewerUser.displayName],
      _rawData: [
        {
          name: 'default',
          rawData: fixturePullRequest,
        },
      ],
    });
  });
});
