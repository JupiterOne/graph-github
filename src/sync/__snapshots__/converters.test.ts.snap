// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toPullRequestEntity with all commits approved 1`] = `
Object {
  "_class": Array [
    "PR",
  ],
  "_key": "JupiterOne/jupiter-integration-aws/pull-requests/45",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "additions": 156,
        "author": Object {
          "isSiteAdmin": true,
          "login": "somebody",
          "name": "Some Body",
        },
        "authorAssociation": "CONTRIBUTOR",
        "baseRefName": "main",
        "baseRefOid": "989c2e36df5b7d2e77ff1ab33aaee49bea705066",
        "baseRepository": Object {
          "name": "jupiter-integration-aws",
          "owner": Object {
            "id": "MDEyOk9yZ2FuaXphdGlvbjQ0NjQ2NTEy",
            "login": "JupiterOne",
            "url": "https://github.com/JupiterOne",
          },
        },
        "body": "This is some discription or something",
        "changedFiles": 7,
        "checksUrl": "https://github.com/JupiterOne/jupiter-integration-aws/pull/45/checks",
        "closed": true,
        "closedAt": "2021-09-01T01:53:34Z",
        "commits": Array [
          Object {
            "author": Object {
              "date": "2021-08-31T18:14:38-04:00",
              "user": Object {
                "isSiteAdmin": true,
                "login": "somebody",
                "name": "Some Body",
              },
            },
            "authoredDate": "2021-08-31T22:14:45Z",
            "changedFiles": 7,
            "commitUrl": "https://github.com/JupiterOne/jupiter-integration-aws/commit/890af24d913107194e9ce9439b7e71ea2b1b17c4",
            "id": "MDY6Q29tbWl0Mzg4ODgwNzEwOjg5MGFmMjRkOTEzMTA3MTk0ZTljZTk0MzliN2U3MWVhMmIxYjE3YzQ=",
            "message": "Ingest AutoScaling Policies and build \`aws_autoscaling_group\` \`USES\` \`aws_autoscaling_policy\` relationships",
            "oid": "890af24d913107194e9ce9439b7e71ea2b1b17c4",
          },
          Object {
            "author": Object {
              "date": "2021-08-31T18:17:39-04:00",
              "user": Object {
                "isSiteAdmin": true,
                "login": "somebody",
                "name": "Some Body",
              },
            },
            "authoredDate": "2021-08-31T22:17:46Z",
            "changedFiles": 0,
            "commitUrl": "https://github.com/JupiterOne/jupiter-integration-aws/commit/69b3d5771a50752acf9310fd93b7a176d6cffaa2",
            "id": "MDY6Q29tbWl0Mzg4ODgwNzEwOjY5YjNkNTc3MWE1MDc1MmFjZjkzMTBmZDkzYjdhMTc2ZDZjZmZhYTI=",
            "message": "Merge branch 'main' of github.com:jupiterone/jupiter-integration-aws into autoscaling-policies",
            "oid": "69b3d5771a50752acf9310fd93b7a176d6cffaa2",
          },
          Object {
            "author": Object {
              "date": "2021-08-31T18:19:34-04:00",
              "user": Object {
                "isSiteAdmin": true,
                "login": "somebody",
                "name": "Some Body",
              },
            },
            "authoredDate": "2021-08-31T22:19:39Z",
            "changedFiles": 2,
            "commitUrl": "https://github.com/JupiterOne/jupiter-integration-aws/commit/1103a7a2e976439ca184f775196759084f16e42f",
            "id": "MDY6Q29tbWl0Mzg4ODgwNzEwOjExMDNhN2EyZTk3NjQzOWNhMTg0Zjc3NTE5Njc1OTA4NGYxNmU0MmY=",
            "message": "fix _type and _class",
            "oid": "1103a7a2e976439ca184f775196759084f16e42f",
          },
        ],
        "createdAt": "2021-08-31T22:15:07Z",
        "deletions": 27,
        "headRefName": "autoscaling-policies",
        "headRefOid": "1103a7a2e976439ca184f775196759084f16e42f",
        "headRepository": Object {
          "name": "jupiter-integration-aws",
          "owner": Object {
            "id": "MDEyOk9yZ2FuaXphdGlvbjQ0NjQ2NTEy",
            "login": "JupiterOne",
            "url": "https://github.com/JupiterOne",
          },
        },
        "id": "MDExOlB1bGxSZXF1ZXN0NzIzODE1Nzc1",
        "isDraft": false,
        "labels": Array [],
        "lastEditedAt": undefined,
        "locked": false,
        "mergeCommit": Object {
          "author": Object {
            "date": "2021-08-31T21:53:34-04:00",
            "user": Object {
              "isSiteAdmin": true,
              "login": "somebody",
              "name": "Some Body",
            },
          },
          "authoredDate": "2021-09-01T01:53:35Z",
          "changedFiles": 7,
          "commitUrl": "https://github.com/JupiterOne/jupiter-integration-aws/commit/39483af8a356b5684513bf56c044dab0ed4c0c7a",
          "id": "MDY6Q29tbWl0Mzg4ODgwNzEwOjM5NDgzYWY4YTM1NmI1Njg0NTEzYmY1NmMwNDRkYWIwZWQ0YzBjN2E=",
          "message": "Merge pull request #45 from JupiterOne/autoscaling-policies",
          "oid": "39483af8a356b5684513bf56c044dab0ed4c0c7a",
        },
        "mergeable": "UNKNOWN",
        "merged": true,
        "mergedAt": "2021-09-01T01:53:34Z",
        "mergedBy": Object {
          "isSiteAdmin": true,
          "login": "somebody",
          "name": "Some Body",
        },
        "number": 45,
        "permalink": "https://github.com/JupiterOne/jupiter-integration-aws/pull/45",
        "publishedAt": "2021-08-31T22:15:07Z",
        "reviewDecision": "APPROVED",
        "reviews": Array [
          Object {
            "author": Object {
              "isSiteAdmin": false,
              "login": "reviewer-user",
              "name": "Reviewer User",
            },
            "commit": Object {
              "oid": "1103a7a2e976439ca184f775196759084f16e42f",
            },
            "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NzQzMjk0Njkx",
            "state": "APPROVED",
            "submittedAt": "2021-09-01T00:19:38Z",
            "updatedAt": "2021-09-01T00:19:38Z",
            "url": "https://github.com/JupiterOne/jupiter-integration-aws/pull/45#pullrequestreview-743294691",
          },
        ],
        "state": "MERGED",
        "title": "Ingest AutoScaling Policies and build relationships",
        "updatedAt": "2021-09-01T01:53:35Z",
        "url": "https://github.com/JupiterOne/jupiter-integration-aws/pull/45",
      },
    },
  ],
  "_type": "github_pullrequest",
  "accountLogin": "JupiterOne",
  "allCommitsApproved": true,
  "approvalLastAt": 1630448379000,
  "approvalsCount": 1,
  "approved": true,
  "approverLogins": Array [
    "reviewer-user",
  ],
  "approvers": Array [
    "Reviewer User",
  ],
  "author": "Some Body",
  "authorLogin": "somebody",
  "commitMessages": Array [
    "Ingest AutoScaling Policies and build \`aws_autoscaling_group\` \`USES\` \`aws_autoscaling_policy\` relationships",
    "Merge branch 'main' of github.com:jupiterone/jupiter-integration-aws into autoscaling-policies",
    "fix _type and _class",
  ],
  "commits": Array [
    "890af24d913107194e9ce9439b7e71ea2b1b17c4",
    "69b3d5771a50752acf9310fd93b7a176d6cffaa2",
    "1103a7a2e976439ca184f775196759084f16e42f",
  ],
  "commitsApproved": Array [
    "890af24d913107194e9ce9439b7e71ea2b1b17c4",
    "69b3d5771a50752acf9310fd93b7a176d6cffaa2",
    "1103a7a2e976439ca184f775196759084f16e42f",
  ],
  "commitsByUnknownAuthor": Array [],
  "commitsCount": 3,
  "commitsNotApproved": Array [],
  "createdOn": 1630448107000,
  "databaseId": undefined,
  "declined": false,
  "description": "This is some discription or something...",
  "displayName": "jupiter-integration-aws/45",
  "id": "45",
  "labels": Array [],
  "mergeCommitHash": "39483af8a356b5684513bf56c044dab0ed4c0c7a",
  "merged": true,
  "mergedOn": 1630461214000,
  "name": "Ingest AutoScaling Policies and build relationships",
  "node": "MDExOlB1bGxSZXF1ZXN0NzIzODE1Nzc1",
  "number": 45,
  "open": false,
  "pullRequestId": "MDExOlB1bGxSZXF1ZXN0NzIzODE1Nzc1",
  "repository": "jupiter-integration-aws",
  "reviewerLogins": Array [
    "reviewer-user",
  ],
  "reviewers": Array [
    "Reviewer User",
  ],
  "source": "autoscaling-policies",
  "state": "MERGED",
  "target": "main",
  "title": "Ingest AutoScaling Policies and build relationships",
  "updatedOn": 1630461215000,
  "validated": true,
  "webLink": "https://github.com/JupiterOne/jupiter-integration-aws/pull/45",
}
`;
